### PR TITLE
some modifications to constant-pH tutorial

### LINKED ISSUE #3211: Check reaction ensemble tutorial with current HEAD

### PR DESCRIPTION
**Pull request by kosovan**
Fixes #3211 

TODO (will do):

- [x] First , explain how the reaction works, then discuss the acceptance

- [x]  Define pH, pK, activity and its relation chemical potential and concentration

- [x]  Refer to original literature for the acceptance, refer to review paper for discussion of peculiarities and pitfalls

- [x] Include all physical parameters of the simulated system (box_l, N0, pKa, pH_range) in one block at the beginning of the script

- [x] Set N0 and concentration in SI units,  calculate box_l in sigma

- [x] restrict the range of pH values to pH in [pK-2:pK+2]

- [x] Remove the discussion of standard error of the mean, refer to Janke instead (statistical analysis is not the learning goal of this tutorial)

- [x] Samples of alpha are correlated. Therefore, apply blocking method or some other analysis of correlated samples. 

- [x] Remove the comparison of histograms, or define a learning goal for this part. Currently, the comparison suggests that constant-pH results differ from theory, while the actual reason for this is insufficient sampling. For a sufficiently long simulations, they should not differ. This is very confusing, especially if it is not explained.

Description of changes:
 - 

PR Checklist
------------
 - [ ] Tests?
   - [ ] Interface
   - [ ] Core 
 - [ ] Docs?

### CODE REVIEW

**Comment by jonaslandsgesell** [comment]
Please go ahead

**Comment by KaiSzuttor** [comment]
can you estimate when this tutorial will be done? We probably want to do 4.1.1 release this week that includes your tutorial.

**Comment by kosovan** [comment]
I am working on it now. I think that the tutorial is quite clean now. 

I suggest that @jonaslandsgesell and @thepith should review the tutorial. 

Still, I would like to add the conversion of units to the SI, and comparison to the reaction ensemble. We can decide later make that part a separate tutorial.

In any case, I want to finish all the work by Thursday.

**Comment by thepith** [comment]
The tutorial is usable as is, only the point
>  Set N0 and concentration in SI units, calculate box_l in sigma

is missing.

In the future we can think of expanding the tutorial to Reaction Ensemble.

**Comment by jonaslandsgesell** [comment]
Thank you. The tutorial looks good to me.
One or two minor things:

- Can you please put the error estimation alogrithm into a function which just returns the estimated error? This would encapsulate the algorithm logically.
I imagine a function which just takes num_As as argument and returns the error.

- Regarding the 95% Confidence interval of the estimated mean value: If the mean value itself is distributed according to a normal distribution, then the 95% CI is given by a z-score of 1.96 (and not 2). Could you link to Wikipedia: https://en.wikipedia.org/wiki/1.96 ? In an ipython notebook you do this via [blue text](url)

PS: Please clear the cells in order to avoid future merge confilicts. (You can do this in the ipython notebook menu: Kernel-> Restart&Clear output, and then saving the notebook)

PPS: maybe you can also run a spell check before final submission?

PPPS: could you also delete the Reaction ensemble tutorial then (the other folder)?

**Comment by kosovan** [comment]
@jonaslandsgesell  @KaiSzuttor 
I am sorry but I cannot do more on this worksheet now. My computer crashed on Thursday, and since then I have been busy with recovering. I resolved some merge conflicts in a text editor but cannot compile espresso and run the notebook on my system. Please, check if it runs cleanly for you before the merge.

**Comment by kosovan** [comment]
I think that it is ready to be merged and used in the tutorials at Espresso school.

**Comment by jonaslandsgesell** [comment]
Can you please merge my fix for the tests:

https://github.com/kosovan/espresso/pull/2

**Comment by jonaslandsgesell** [comment]
The tutorial tests still fail. See CI.

**Comment by jonaslandsgesell** [comment]
Please try to merge my new PR, it might fix the test case again

File: doc/tutorials/12-constant_pH/si_units.py
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/si_units.py]: if this file should be part of espresso in the future, we should not pollute the namespace with such global variables that are just aliases.

File: doc/tutorials/12-constant_pH/si_units.py
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/si_units.py]: please adhere to the naming convention, namely ```CamelCase``` for classes.

File: doc/tutorials/12-constant_pH/si_units.py
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/si_units.py]: instead of ```N_Av``` just explicitly use ```scipy.constatns.N_A```

File: doc/tutorials/12-constant_pH/si_units.py
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/si_units.py]: ```amu``` is actually newer used anywhere

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]

File: doc/tutorials/12-constant_pH/si_units.py
Comment (kosovan) [reply doc/tutorials/12-constant_pH/si_units.py]: @KaiSzuttor would you like it to be calles SIunits or SiUnits? I vote for the first option because it reads easier, although it is not quite camelcase.

File: doc/tutorials/12-constant_pH/si_units.py
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/si_units.py]: actually there already exists a package that may be doing what you are looking for:
https://pint.readthedocs.io/

Comment (KaiSzuttor) [reply doc/tutorials/12-constant_pH/si_units.py]: *e.g.*:
``` python
#!/usr/bin/env python3

import scipy.constants
import pint

ureg = pint.UnitRegistry()

sigma = 5 * ureg.nanometer
tau = 1.0 * ureg.picosecond

c = 0.1 * ureg.mol / ureg.litre
box_volume = 10 * sigma * 20 * sigma * 5 * sigma
N_part = int((c * box_volume).to(ureg.mole).magnitude * scipy.constants.N_A)
```

File: doc/tutorials/12-constant_pH/si_units.py
Comment (jonaslandsgesell) [inline doc/tutorials/12-constant_pH/si_units.py]: Does this now return that in order to convert from 1/sigma^3 (with sigma=0.355nm) to mol/l, I need to multiply with roughly 37.1? Just checking for a result I know...

Comment (kosovan) [reply doc/tutorials/12-constant_pH/si_units.py]: To get the answer, just add the following line anywhere after initialization of the SI_units class:
`print(SI.convert(1.0,'N/sigma^3','mol/L'))`

Result: 37.11624488606752

Comment (jonaslandsgesell) [reply doc/tutorials/12-constant_pH/si_units.py]: Great! That s easy.

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (jonaslandsgesell) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: I don't think this is needed. You can run MD and MC moves sequentially. This is normal practice. Sequentiality will only affect how fast one converges to equilibrium. In the limit of infinite runtime the exact sequence of MC/MD moves does not matter anymore. https://link.springer.com/chapter/10.1007/978-3-642-41315-5_23 See figure 1.

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (jonaslandsgesell) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: You should make using the integrator conditional on USE_WCA in order to speed up the ideal case:

```
        else:
            if USE_WCA==True:
                system.integrator.run(steps=1000);
```

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (kosovan) [reply doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: It is common practice but I am not aware of any proof that it is correct. In general, performing MC moves sequentially instead of a random order breaks detailed balance. If there is enough randomness in the system, this breaking might be weak and hardly noticeable. Sot it is not advisable to do it like that.

Even if we routinely break this rule, we should not teach others to do so.

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (kosovan) [reply doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: Yes, we can do it.

File: doc/tutorials/12-constant_pH/si_units.py
Comment (kosovan) [reply doc/tutorials/12-constant_pH/si_units.py]: I will soon commit a different approach, that uses the pint package and internally checks for correct dimensionality.

File: doc/tutorials/12-constant_pH/si_units.py
Comment (kosovan) [reply doc/tutorials/12-constant_pH/si_units.py]: DONE

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (jonaslandsgesell) [reply doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: I am not aware of a proof on how detailed balance is broken by sequential MC/MD moves. Are you?

File: doc/tutorials/12-constant_pH/si_units.py
Comment (kosovan) [reply doc/tutorials/12-constant_pH/si_units.py]: fixed

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (kosovan) [reply doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: Some years back, Olaf Lenz was showing an example of that at the ICP seminar.  I cannot immediately reproduce it now.

Are you aware that a random order could break detailed balance? If not, then we should keep the random order as a safe choice.

**Comment by kosovan** [comment]
@jonaslandsgesell @thepith Please, have a look at my last modifications regarding the SI units. If you do not find an error, then I think this tutorial is ready for a merge.

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (jonaslandsgesell) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: This is a desired change. However this breaks the tests of the tutorials. Please also update the sphinx documentation regarding the installation requirements.

@KaiSzuttor  we need dockerfiles with pint installed. How do we proceed here. Update the Docker images?

**Review by jonaslandsgesell — CHANGES_REQUESTED** [review]
The test cases need to be fixed. This requires Docker images to be updated to include pint.

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (jonaslandsgesell) [reply doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: I do not know published literature that prohibits sequentially using MD and MC moves.

The paper "A rigorous procedure for combining molecular dynamics and Monte Carlo simulation algorithms" demonstrates that the correct probability distribution is sampled for the canonical (NVT) case:
https://www.sciencedirect.com/science/article/pii/S0301010400002469

They perform MDMC steps. One MDMC step consists of 

1. m MC steps
2. integrating n timesteps with MD

In section 3 they show that this is a valid approach.

Comment (thepith) [reply doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: The problem is that you are proposing a sequential MC scheme. While this can be correct, it can also fail. See the discussion in [1]:

> The implications of our theorem are significant. First, the sequential updating approach, widely practiced and nearly as widely criticized, is correct. All that is required is that the
sweep leave the Boltzmann distribution invariant, in addition to leading to regular sampling. Indeed, there are numerical reports that the sequential updating approach leads not only
to correct sampling but also to more efficient sampling in a variety of contexts, ranging from image processing to membrane physics. There are some simple cases, however, for
which random Metropolis updating is ergodic and regular, but sequential Metropolis updating is not. One is the Ising model at infinite temperature, where the configurations produced by sequential updating simply oscillate between the initial configuration and the configuration with all spins flipped. Another example is the 2x2 Ising model at any temperature. If the spins are updated in the order [[1 2][3 4]], then the initial configurations [[+ -][+ +]] and [[+ +][- +]] lead to sequences that also oscillate between spin–flipped configurations. In these examples, our theorem does not guarantee convergence to the Boltzmann distribution, because assumption (4) is violated. If, however, the individual spin–flip moves within the sweep were performed with a probability (1 - a), as in Eq. (5), then convergence to the Boltzmann distribution would occur in both of these examples.

This shows that sequential updating can break the MC simulation. Therefore, we should be careful how we teach it.

The paper you cite they show that you can combine configurational MC with MD. But they do not discuss whether this is true when you combine "orthogonal" algorithms like cpH and MD (where one samples the configurational space and the other the composition space).

It might well be that the sequential updating of the cpH and the MD move leads to correct sampling. But what is the advantage to do this in this tutorial? I do not see any real advantage, but the big disadvantage that we teach an advanced method which is not generally valid. If we want to use the sequential updating, then we should discuss why we want use it, why it works and when does not work. All of this is out of scope of this tutorial.

The random mixing of the moves is the most basic way how to do it, and in this tutorial we should stick to be basics when it comes to such subjects.

[1] Manousiouthakis, V. I.; Deem, M. W. Strict Detailed Balance Is Unnecessary in Monte Carlo Simulation. The Journal of Chemical Physics 1999, 110 (6), 2753–2756. https://doi.org/10.1063/1.477973.

**Comment by kosovan** [comment]
@jonaslandsgesell can you please check our latest changes? 

In my opinion, the tutorial is now ready to be merged. 

There is still a pending review from @KaiSzuttor

**Comment by jngrad** [comment]
The tutorial test fails on my machine due to the use of unknown variable `degrees_of_dissociation`.

**Comment by kosovan** [comment]
> The tutorial test fails on my machine due to the use of unknown variable `degrees_of_dissociation`.

Because the file `test.py` still runs an old version of the tutorial. I will fix that.

Update: Sorry, I initially misunderstood what was going on. This file existed only locally on my computer. Now I think I know what the trouble is, and will try to fix it.

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (jngrad) [reply doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: `pint` was installed in docker and the requirements updated in 4d655b8

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (jonaslandsgesell) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: This is not quite true. This depends on the observable you are looking at. In an ideal system the pressure is given by P=\sum_i c_i kT. If the concentration of some species i is wrongly represented, then the pressure which is observed is already wrong. This can be significant in the cpH ensemble at low or high pH.  Please adapt the sentence to reflect this fact.

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (jonaslandsgesell) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: Please drop the comments they easily get outdated.

**Comment by jonaslandsgesell** [comment]
The new figures are quite helpful in understanding the limitations of the cpH ensemble. Thank you for contributing those.

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (jonaslandsgesell) [reply doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: I would drop the comment, or say "module for working with units and dimensions"

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (kosovan) [reply doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: @jonaslandsgesell I do not understand. What could become outdated in these comments?

Comment (jonaslandsgesell) [reply doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: the information that av_c is in mol/l is redundant, because this is carried over from c_acid. This can become outdated if someone changes the unit of c_acid.
You might consider variable names which are more self-explaining. Then the comment is also not needed anymore.

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (kosovan) [reply doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: This is a more general issue, namely what is the meaning of "pressure" in implicit solvent simulations. One could claim that it is the osmotic pressure, but then one needs to specify what is on the other side of the osmotic membrane. This discussion is for sure outside the scope of the reaction toturial

I can address the comment by explicitly stating that the ions do not affect the ionization degree.

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (kosovan) [reply doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: I will make the variable names more explicit.

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (kosovan) [reply doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: ok

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (jonaslandsgesell) [reply doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: The instantaneous "kinetic" virial pressure tensor component i,j is given by:
image,
The last line is the ensemble averaged kinetic virial pressure tensor component i,j. Via some lines of math using the Maxwell-Boltzmann distribution of velocity components one obtains <P_i,j>=\sum_k c_k KT (where k runs over all ion species).

The instantaneous kinetic virial pressure tensor is returned by the pressure tensor command in Espresso.
The result of this observable (and also the pressure observable) will be affected if there are e.g. H+ ions missing at low pH in the cpH ensemble. Therefore the statement  "the ions missing in the simulation have no effect on the results [in an ideal system]" does not hold in general.

**Comment by jonaslandsgesell** [comment]
This is fine for me to merge. Thank you for your improvements.

**Comment by kosovan** [comment]
@jonaslandsgesell @KaiSzuttor 
I think that this PR can be merged now.
Github says that there is one pending change request from @KaiSzuttor but I cannot find it. Can anybody please point me to what is still missing?

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: could you please check the coding style of this part? Especially:
* please use spaces in assignments: ```a = 3```
* please one assignment per line ~~```a=3; b=2```~~
* is the global system seed needed her? I thought only NPT is left using a global seed...

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: please move all imports to the top

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: I think if you can come up with more expressive variable names it would help to make it more readable:
*e.g.*: 
``` python
for p in polymer:
    for index, monomer_pos in enumerate(p):
```

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: this can probably simplified to ```print(p3m.get_params())```

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: the comment should explain why there are integrations before setting up electrostatics

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: please check the naming convention for variables: ```long_variable_name``` instead of ```longvariablename```

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: if this is a constant please use capital letters for the name

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: constant?

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: please remove the commented out code

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: please come up with a more expressive variable name which maybe makes it possible to get rid of the comment:
*e.g.* ```num_of_A```?

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: have you checked that the citation works?

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: one of the comments needs to go

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (KaiSzuttor) [inline doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: please also check this code for formatting and style

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]
i've added some comments on the code... I think it's not a lot of work. If my comments are addressed this can be merged.

**Comment by RudolfWeeber** [comment]
https://jupyter-contrib-nbextensions.readthedocs.io/en/latest/nbextensions/code_prettify/README_autopep8.html

File: doc/tutorials/12-constant_pH/12-constant_pH.ipynb
Comment (kosovan) [reply doc/tutorials/12-constant_pH/12-constant_pH.ipynb]: I made the variable name more explicit but still prefer to keep the comment.

**Comment by kosovan** [comment]
@KaiSzuttor I hope now it is ready.

**Comment by KaiSzuttor** [comment]
thanks @kosovan  I think it's fine now, there are still some inconsistencies with respect to style but I think we have to check for some tools to automatically do this. There are plugins for jupyter to apply such style changes automatically. I'll try it and will push to this PR.

**Comment by KaiSzuttor** [comment]
@kosovan please check again, I think I only touched style plus removed the output

**Comment by kosovan** [comment]
I think that everything is OK.

**Review by KaiSzuttor — APPROVED** [review]

**Comment by kosovan** [comment]
> -1 Rejected by code reviews

Unfortunately, I do not understand what this means .

**Comment by KaiSzuttor** [comment]
> Unfortunately, I do not understand what this means .

never mind that's our merge-bot that did not want to merge because CI was not green yet.

**Review by jonaslandsgesell — APPROVED** [review]

**Comment by jngrad** [comment]
@KaiSzuttor what is bors merging right now? There are two CI pipelines for `staging` currently running: this PR merge and yesterday's merge, the latter having replaced this PR merge in the `staging` history.

**Pull request merged**
